src/server-logger.ts

```diff
@@ -0,0 +1,257 @@
+      bufferItem = this.logBuffer[this.logBuffer.length - 1];
+    } else {
+      bufferItem = {
+        size: 0,
+        traces: []
+      };
+    }
+
+    /* if pushing our trace onto the buffer item will be too large, we need a new buffer item */
+    const exceedsMaxLogSize = bufferItem.size + traceMessageSize > MAX_LOG_SIZE;
+    if (exceedsMaxLogSize) {
+      // console.log('%c `exceedsMaxLogSize` was `true`', 'color: #32a852', { logBuffer: [...this.logBuffer], bufferItem, incomingTrace: trace, incomingTraceSize: traceMessageSize }); // TODO: delete me
+      this.logBuffer.push({
+        size: traceMessageSize,
+        traces: [trace]
+      });
+      /* since we pushed a new item, we need to send immediately */
+      // console.log('%c calling sendLogsToServer(true)', 'color: #32a852', { logBuffer: [...this.logBuffer] }); // TODO: delete me
+      this.sendLogsToServer(true);
+      return;
+    }
+
+    /* else just push onto the buffer */
+    bufferItem.size += traceMessageSize;
+    bufferItem.traces.push(trace);
+
+    /* if we don't have anything in the buffer, that means we have to push this new item */
+    if (!this.logBuffer.length) {
+      // console.log('%c `this.logBuffer` was empty. pushing new buffer item', 'color: #32a852', { logBuffer: [...this.logBuffer], bufferItem }); // TODO: delete me
+      this.logBuffer.push(bufferItem);
+    }
+
+    /* this will setup the debounce timer (if it is not already running) */
+    // console.log('%c calling sendLogsToServer()', 'color: #32a852', { logBuffer: [...this.logBuffer] }); // TODO: delete me
+    this.sendLogsToServer();
+  }
+
+  private async sendLogsToServer (immediate: boolean = false): Promise<any> {
+    if (!this.logBuffer.length) {
+      /* clear timer */
+      clearTimeout(this.debounceTimer);
+      this.debounceTimer = null;
+      // console.log('%c buffer empty, not sending http request', 'color: #32a852'); // TODO: delete me
+      return;
+    }
+
+    /* if we have a pending request, then don't send another */
+    if (this.pendingHttpRequest) {
+      /* clear timer */
+      clearTimeout(this.debounceTimer);
+      this.debounceTimer = null;
+      // console.log('%c http request pending, not sending another http request', 'color: #32a852'); // TODO: delete me
+      return;
+    }
+
+    /* if we aren't sending immediately, then setup the timer */
+    if (!immediate) {
+      // console.log('%c `immediate` is false. setting up debounce timer', 'color: #32a852'); // TODO: delete me
+      if (!this.debounceTimer) {
+        // console.log('%c setting `debounceTimer`', 'color: #32a852'); // TODO: delete me
+        this.debounceTimer = setTimeout(() => this.sendLogsToServer(true), this.debounceLogUploadTime);
+      }
+      return;
+    }
+
+    /* clear timer */
+    clearTimeout(this.debounceTimer);
+    this.debounceTimer = null;
+
+    /* grab the first item to send */
+    const bufferItem = this.logBuffer[0];
```

Comment:
what are your thoughts on doing something like `const bufferItem = this.logBuffer.shift()` to remove it automatically, then doing a `unshift` in the catch block to move it back to the front of the array. I think this approach may be safer in case we end up here again before the bufferItem is actually removed.

Reply:
That would also simplify the logic in `sendAllLogsInstantly`

Reply:
This is a good approach. I didn't think about the possibility of adding more traces to a logBuffer item that was already in flight.